drivers/led/lp50xx.c

```diff
@@ -0,0 +1,282 @@
+lp50xx_led_to_info(const struct lp50xx_config *config, uint32_t led)
+{
+	for (uint8_t i = 0; i < config->num_leds; i++) {
+		if (config->leds_info[i].index == led) {
+			return &config->leds_info[i];
+		}
+	}
+	return NULL;
+}
+
+int lp50xx_get_info(const struct device *dev, uint32_t led, const struct led_info **info)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int lp50xx_set_brightness(const struct device *dev,
+				 uint32_t led, uint8_t value)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[2];
+
+	if (!led_info || value > 100) {
+		return -EINVAL;
+	}
+
+	buf[0] = LP50XX_LED_BRIGHTNESS_BASE(config->chip) + led_info->index;
+	buf[1] = (value * 0xff) / 100;
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_on(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 100);
+}
+
+static int lp50xx_off(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 0);
+}
+
+static int lp50xx_set_color(const struct device *dev, uint32_t led,
+			    uint8_t num_colors, const uint8_t *color)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[4];
+
+	if (!led_info || num_colors != led_info->num_colors) {
+		return -EINVAL;
+	}
+
+	buf[0] = LP50XX_OUT_COLOR_BASE(config->chip) + 3 * led_info->index;
+	buf[1] = color[0];
+	buf[2] = color[1];
+	buf[3] = color[2];
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_write_channels(const struct device *dev,
+				 uint32_t start_channel,
+				 uint32_t num_channels, const uint8_t *buf)
+{
+	const struct lp50xx_config *config = dev->config;
+	struct lp50xx_data *data = dev->data;
+
+	if (start_channel >= LP50XX_NUM_CHANNELS(config->chip) ||
+	    start_channel + num_channels > LP50XX_NUM_CHANNELS(config->chip)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Unfortunately this controller don't support commands split into
+	 * two I2C messages.
+	 */
+	data->chan_buf[0] = (config->chip) + start_channel;
+	memcpy(data->chan_buf + 1, buf, num_channels);
+
+	return i2c_write_dt(&config->bus, data->chan_buf, num_channels + 1);
+}
+
+static int lp50xx_init(const struct device *dev)
+{
+	const struct lp50xx_config *config = dev->config;
+	uint8_t buf[3];
+	int err;
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	if (config->num_leds > LP50XX_MAX_LEDS(config->chip)) {
+		LOG_ERR("%s: invalid number of LEDs %d (max %d)",
+			dev->name, config->num_leds, LP50XX_MAX_LEDS(config->chip));
+		return -EINVAL;
+	}
+
+	/* Enable EN pin if given in devicetree. */
+	if (device_is_ready(config->en_pin.port)) {
```

Comment:
As en_gpios is set to `required: false` in ti,lp50xx.yaml file I think it's better to just do nothing?

Reply:
Well,  if the GPIO is actually passed, then it **must** be configured, else the controller won't work. Correct ? So if we can't because the GPIO chip is not ready, then I think we have to log the error. But indeed if `device_is_ready` returns false because the GPIO is not passed, then we don't want to log anything.